{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "breathing-external",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################\n",
    "#####Figure S3C ######\n",
    "######################\n",
    "#read in CLIP samfile\n",
    "#divide samfile by chromosome (using samtools)?\n",
    "\n",
    "#parse through fasta file for each chromosome and compare with CLIP tags\n",
    "#calculate shannon entropy scores \n",
    "#filter tags with high entropy scores\n",
    "\n",
    "#question: calculate shannon entropy scores along each chromosome position or along tags???\n",
    "\n",
    "#find the position with the highest entropy scores and re-align tags \n",
    "\n",
    "\n",
    "#LIN28A binding motif identified by CLIP-seq. \n",
    "#Sequences harboring a mutation were aligned with the mutated nucleotide centered at zero. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excess-revision",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "tight-cannon",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import pysam\n",
    "\n",
    "def readinSAM(chrNum):\n",
    "    samfile = pysam.AlignmentFile(\"./LIN28A_CLIP.sorted.bam\", \"rb\")\n",
    "    cigarDic = defaultdict(str)\n",
    "    seqDic = defaultdict(str)\n",
    "    mapDic = defaultdict(int)\n",
    "    reverseDic = defaultdict(bool)\n",
    "    for read in samfile.fetch(chrNum):\n",
    "        #if read.is_reverse == False:\n",
    "        index = read.query_name\n",
    "        cigarDic[index] = read.cigarstring\n",
    "        seqDic[index] = read.query_sequence\n",
    "        mapDic[index] = read.reference_start\n",
    "        reverseDic[index] = read.is_reverse\n",
    "    samfile.close()\n",
    "    return cigarDic, seqDic, mapDic, reverseDic\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "decent-summit",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get fasta sequence for each chromosome\n",
    "from Bio import SeqIO\n",
    "\n",
    "def readinFasta(chrNum):\n",
    "    for record in SeqIO.parse(\"./references/GRCm39.primary_assembly.genome.fa\", \"fasta\"):\n",
    "        if record.id == chrNum:\n",
    "            refseq = str(record.seq)\n",
    "    return refseq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "reverse-interstate",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import re\n",
    "softclip = re.compile('[S](\\d)*')\n",
    "match = re.compile('(\\d)*[M]')\n",
    "deletion = re.compile('[D](\\d)*')\n",
    "insertion = re.compile('[I](\\d)*')\n",
    "forsplit = re.compile('(\\d*[A-Z])')\n",
    "\n",
    "def matchFrame (cigarDic, seqDic):\n",
    "    newseqDic = defaultdict(str)\n",
    "    for read in seqDic.keys():\n",
    "        cigar = cigarDic[read]\n",
    "        sequence = seqDic[read]\n",
    "        newseq = ''\n",
    "        count = 0\n",
    "        cigarlist = forsplit.split(cigar)\n",
    "        for i in cigarlist:\n",
    "            if match.search(i) != None:\n",
    "                matched =  int(i.rstrip('M'))\n",
    "                newseq += sequence[count:count+matched]\n",
    "                count += matched\n",
    "            elif softclip.search(i) != None:\n",
    "                clippedLen = int(i.rstrip('S'))\n",
    "                count += clippedLen\n",
    "            elif deletion.search(i) != None:\n",
    "                deletedLen = int(i.rstrip('D'))\n",
    "                newseq += 'N'*deletedLen\n",
    "            elif insertion.search(i) != None:\n",
    "                insertedLen = int(i.rstrip('I'))\n",
    "                count += insertedLen\n",
    "                sequence = sequence[:count]+sequence[count+insertedLen:]\n",
    "                newseq += sequence\n",
    "        newseqDic[read] = newseq\n",
    "    return dict(sorted(newseqDic.items()))\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "eight-relief",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TTTT')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "a = Seq('ACTG')\n",
    "a.reverse_complement()\n",
    "Seq('AAAA').reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "minute-tokyo",
   "metadata": {},
   "outputs": [],
   "source": [
    "#go through each frame-matched read sequence and compare with the ref sequence\n",
    "#create numpy array for calculating frequencies\n",
    "#A, T, C, G, N\n",
    "\n",
    "#calculate shannon entropy at once?\n",
    "\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "import datetime\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "nuc_dict = {'A':0, 'C':1, 'G':2, 'T':3, 'N':4}\n",
    "\n",
    "def seqcompare (chrseq, edited_seqDic, chr_mapDic, reverseDic):\n",
    "    pos_nucfreq_arr = np.zeros((len(chrseq), 5), dtype=float)\n",
    "    neg_nucfreq_arr = np.zeros((len(chrseq), 5), dtype=float)\n",
    "    pos_depth_arr = np.zeros((len(chrseq), 1), dtype=int)\n",
    "    neg_depth_arr = np.zeros((len(chrseq), 1), dtype=int)\n",
    "    \n",
    "    for count, read in enumerate(edited_seqDic.keys()):\n",
    "        if count%1000000==0: \n",
    "            print(datetime.datetime.now())\n",
    "        clip_seq = edited_seqDic[read]\n",
    "        pos = chr_mapDic[read]\n",
    "        for nuc in clip_seq:\n",
    "            if reverseDic[read] == False:\n",
    "                pos_depth_arr[pos,0] += 1\n",
    "                if chrseq[pos] != nuc:\n",
    "                    pos_nucfreq_arr[pos, nuc_dict[nuc]] += 1\n",
    "                pos += 1\n",
    "                if pos == len(chrseq):\n",
    "                    break\n",
    "            elif reverseDic[read] == True:\n",
    "                neg_depth_arr[pos,0] += 1\n",
    "                if chrseq[pos] != nuc:\n",
    "                    neg_nucfreq_arr[pos, nuc_dict[nuc]] += 1\n",
    "                pos += 1\n",
    "                if pos == len(chrseq):\n",
    "                    break\n",
    "            \n",
    "    return pos_nucfreq_arr, neg_nucfreq_arr, pos_depth_arr, neg_depth_arr\n",
    "                \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "compound-national",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate shannon entropy for each position\n",
    "\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "from scipy.special import entr\n",
    "\n",
    "\n",
    "def calcShannon (nuc_arr):\n",
    "    with np.errstate(divide='ignore', invalid='ignore'):\n",
    "        prob_arr = nuc_arr/(nuc_arr.sum(axis=1, keepdims=True)+10e-20)\n",
    "    e_arr = entr(prob_arr).sum(axis=1, keepdims=True)\n",
    "    return e_arr\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorrect-passport",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "motivated-pearl",
   "metadata": {},
   "outputs": [],
   "source": [
    "#retrieve list of chromosomes\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def getChrlist():\n",
    "    chr_list = []\n",
    "    for record in SeqIO.parse(\"./references/GRCm39.primary_assembly.genome.fa\", \"fasta\"):\n",
    "        if record.id not in chr_list:\n",
    "            chr_list.append(record.id)\n",
    "    return chr_list\n",
    "\n",
    "chr_list = getChrlist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "billion-rwanda",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "curious-presentation",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter each nucleotide position by its depth and entropy\n",
    "def filter_index_list(entropy_arr, depth_arr):\n",
    "    np.where(depth_arr > 50, entropy_arr, 0) #change entropy values in entropy_arr to 0 if smaller than 50\n",
    "    filtered_ind = np.where(entropy_arr > 0.8) #index of entropy values bigger than 0.8\n",
    "    filtered_entropy = np.take(entropy_arr, filtered_ind)[0,:] #get entropy values in accordance with filtered index values\n",
    "    temp_idx = (0,0)\n",
    "    temp_ind = 0\n",
    "    temp_e = 0\n",
    "    ind_list = list()\n",
    "    e_list = list()\n",
    "    for idx, ind in np.ndenumerate(filtered_ind[0]):\n",
    "        #if ind == 48689527 or ind == 48689528:\n",
    "         #   print('correct')\n",
    "          #  print(filtered_entropy[idx])\n",
    "        if ind-temp_ind <= 10: #compare two nucleotide sites if they are close (<10nt)\n",
    "            if filtered_entropy[idx] >= filtered_entropy[temp_idx]:\n",
    "                del ind_list[-1]\n",
    "                ind_list.append(ind)\n",
    "                del e_list[-1]\n",
    "                e_list.append(filtered_entropy[idx])\n",
    "        else: #add new index and entropy to list if it's far away from previously saved data\n",
    "            ind_list.append(ind)\n",
    "            e_list.append(filtered_entropy[idx])\n",
    "        temp_ind = ind\n",
    "        temp_idx = idx\n",
    "        temp_e = filtered_entropy[idx]\n",
    "    ind_array = np.array(ind_list)\n",
    "    e_array = np.array(e_list)\n",
    "    return ind_array, e_array\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "welsh-chemical",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "missing-encoding",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "narrative-healthcare",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-02 12:33:36.063112\n",
      "2021-07-02 12:33:53.756122\n",
      "2021-07-02 12:34:10.694829\n",
      "end of chr1\n",
      "2021-07-02 12:35:42.951719\n",
      "2021-07-02 12:36:00.635008\n",
      "2021-07-02 12:36:17.803591\n",
      "end of chr2\n",
      "2021-07-02 12:37:28.745389\n",
      "2021-07-02 12:37:45.511649\n",
      "end of chr3\n",
      "2021-07-02 12:38:59.071595\n",
      "2021-07-02 12:39:16.027080\n",
      "end of chr4\n",
      "2021-07-02 12:40:29.148226\n",
      "2021-07-02 12:40:46.460482\n",
      "end of chr5\n",
      "2021-07-02 12:42:00.208138\n",
      "2021-07-02 12:42:17.777245\n",
      "2021-07-02 12:42:35.193308\n",
      "end of chr6\n",
      "2021-07-02 12:43:31.365152\n",
      "2021-07-02 12:43:48.249737\n",
      "end of chr7\n",
      "2021-07-02 12:44:54.667182\n",
      "2021-07-02 12:45:13.199511\n",
      "end of chr8\n",
      "2021-07-02 12:46:21.321153\n",
      "2021-07-02 12:46:38.549374\n",
      "end of chr9\n",
      "2021-07-02 12:47:45.440188\n",
      "2021-07-02 12:48:02.813876\n",
      "end of chr10\n",
      "2021-07-02 12:49:17.811425\n",
      "2021-07-02 12:49:35.936393\n",
      "2021-07-02 12:49:53.627579\n",
      "end of chr11\n",
      "2021-07-02 12:50:57.134396\n",
      "2021-07-02 12:51:13.917766\n",
      "end of chr12\n",
      "2021-07-02 12:52:10.252982\n",
      "2021-07-02 12:52:27.718171\n",
      "end of chr13\n",
      "2021-07-02 12:53:24.668903\n",
      "2021-07-02 12:53:41.935855\n",
      "end of chr14\n",
      "2021-07-02 12:54:36.886384\n",
      "2021-07-02 12:54:54.590470\n",
      "end of chr15\n",
      "2021-07-02 12:55:49.459054\n",
      "2021-07-02 12:56:07.093644\n",
      "end of chr16\n",
      "2021-07-02 12:57:49.113863\n",
      "2021-07-02 12:58:09.716243\n",
      "2021-07-02 12:58:29.770155\n",
      "2021-07-02 12:58:49.621879\n",
      "2021-07-02 12:59:09.361629\n",
      "2021-07-02 12:59:29.071782\n",
      "end of chr17\n",
      "2021-07-02 13:00:33.097576\n",
      "2021-07-02 13:00:50.279669\n",
      "end of chr18\n",
      "2021-07-02 13:01:33.774014\n",
      "2021-07-02 13:01:51.246726\n",
      "end of chr19\n",
      "2021-07-02 13:02:30.406966\n",
      "2021-07-02 13:02:46.752446\n",
      "end of chrX\n",
      "2021-07-02 13:03:39.517555\n",
      "end of chrY\n",
      "2021-07-02 13:04:09.358910\n",
      "end of chrM\n",
      "2021-07-02 13:04:24.777818\n",
      "end of GL456210.1\n",
      "2021-07-02 13:04:39.421176\n",
      "end of GL456211.1\n",
      "2021-07-02 13:04:54.290261\n",
      "end of GL456212.1\n",
      "2021-07-02 13:05:08.986657\n",
      "end of GL456219.1\n",
      "2021-07-02 13:05:23.897785\n",
      "end of GL456221.1\n",
      "2021-07-02 13:05:38.653214\n",
      "end of GL456233.2\n",
      "2021-07-02 13:05:53.630298\n",
      "end of GL456239.1\n",
      "2021-07-02 13:06:08.420780\n",
      "end of GL456354.1\n",
      "2021-07-02 13:06:23.191945\n",
      "end of GL456359.1\n",
      "2021-07-02 13:06:37.938409\n",
      "end of GL456360.1\n",
      "2021-07-02 13:06:52.631127\n",
      "end of GL456366.1\n",
      "2021-07-02 13:07:07.423706\n",
      "end of GL456367.1\n",
      "2021-07-02 13:07:22.196799\n",
      "end of GL456368.1\n",
      "2021-07-02 13:07:37.223423\n",
      "end of GL456370.1\n",
      "2021-07-02 13:07:52.056255\n",
      "end of GL456372.1\n",
      "2021-07-02 13:08:06.858085\n",
      "end of GL456378.1\n",
      "2021-07-02 13:08:21.585605\n",
      "end of GL456379.1\n",
      "2021-07-02 13:08:36.368221\n",
      "end of GL456381.1\n",
      "2021-07-02 13:08:51.275196\n",
      "end of GL456382.1\n",
      "2021-07-02 13:09:06.647412\n",
      "end of GL456383.1\n",
      "2021-07-02 13:09:21.301992\n",
      "end of GL456385.1\n",
      "2021-07-02 13:09:36.059317\n",
      "end of GL456387.1\n",
      "2021-07-02 13:09:50.792646\n",
      "end of GL456389.1\n",
      "2021-07-02 13:10:05.561620\n",
      "end of GL456390.1\n",
      "2021-07-02 13:10:20.309734\n",
      "end of GL456392.1\n",
      "2021-07-02 13:10:35.013908\n",
      "end of GL456394.1\n",
      "2021-07-02 13:10:49.777274\n",
      "end of GL456396.1\n",
      "2021-07-02 13:11:04.510319\n",
      "end of JH584295.1\n",
      "2021-07-02 13:11:19.268539\n",
      "end of JH584296.1\n",
      "2021-07-02 13:11:34.111962\n",
      "end of JH584297.1\n",
      "2021-07-02 13:11:48.868249\n",
      "end of JH584298.1\n",
      "2021-07-02 13:12:03.561270\n",
      "end of JH584299.1\n",
      "2021-07-02 13:12:18.397079\n",
      "end of JH584300.1\n",
      "2021-07-02 13:12:33.228807\n",
      "end of JH584301.1\n",
      "2021-07-02 13:12:48.033021\n",
      "end of JH584302.1\n",
      "2021-07-02 13:13:02.850330\n",
      "end of JH584303.1\n",
      "2021-07-02 13:13:17.736881\n",
      "end of JH584304.1\n",
      "2021-07-02 13:13:32.608429\n",
      "end of MU069434.1\n",
      "2021-07-02 13:13:49.048107\n",
      "end of MU069435.1\n"
     ]
    }
   ],
   "source": [
    "#repeat all of the process for all chromosomes\n",
    "total_idx_arr = np.array([])\n",
    "total_entr_arr = np.array([])\n",
    "chr_mark_list = list()\n",
    "\n",
    "\n",
    "#chr_list = {'chr11'}\n",
    "for chr in chr_list:\n",
    "    chr_cigarDic, chr_seqDic, chr_mapDic, chr_reverseDic = readinSAM(chr)\n",
    "    chrseq = readinFasta(chr)\n",
    "    edited_chr_seqDic = matchFrame(chr_cigarDic, chr_seqDic)\n",
    "    pos_chr_nuc_arr, neg_chr_nuc_arr, pos_chr_depth_arr, neg_chr_depth_arr = seqcompare(chrseq, edited_chr_seqDic, chr_mapDic, chr_reverseDic)   \n",
    "    pos_entropy_arr = calcShannon(pos_chr_nuc_arr)\n",
    "    neg_entropy_arr = calcShannon(neg_chr_nuc_arr)\n",
    "    pos_idx_arr, pos_entr_arr = filter_index_list(pos_entropy_arr, pos_chr_depth_arr)\n",
    "    neg_idx_arr, neg_entr_arr = filter_index_list(neg_entropy_arr, neg_chr_depth_arr)\n",
    "    chr_mark_list.extend(['p'+chr]*len(pos_idx_arr))\n",
    "    chr_mark_list.extend(['n'+chr]*len(neg_idx_arr))\n",
    "    total_idx_arr = np.append(total_idx_arr, pos_idx_arr)\n",
    "    total_idx_arr = np.append(total_idx_arr, neg_idx_arr)\n",
    "    total_entr_arr = np.append(total_entr_arr, pos_entr_arr)\n",
    "    total_entr_arr = np.append(total_entr_arr, neg_entr_arr)   \n",
    "    print(\"end of %s\"%chr)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "cardiovascular-submission",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29150"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(total_idx_arr)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "hawaiian-johnson",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rank each positions by entropy and keep only nucleotide positions with the highest 1000 entropies\n",
    "ind = np.argpartition(total_entr_arr, -1000)[-1000:] #get index for top 1000 entropy in entropy array\n",
    "complete_pos_arr = np.take(total_idx_arr, ind).astype(int) #keep nucleotide values only in accordance with the previously retrieved index\n",
    "top_chr_mark_list = list()\n",
    "for i in ind:\n",
    "    top_chr_mark_list.append(chr_mark_list[i]) #get chromosome values only in accordance with the previously retrieved index \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lovely-bouquet",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "isolated-genre",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "peaceful-spain",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "major-packing",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get fasta sequence for each mapped position and write sequence as in a new fasta file\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def writeFasta (chrNum, chr_pos_list, strand_list):\n",
    "    for record in SeqIO.parse(\"./references/GRCm39.primary_assembly.genome.fa\", \"fasta\"):\n",
    "        if record.id == chrNum:\n",
    "            for i in range(len(chr_pos_list)):\n",
    "                records = []\n",
    "                pos = chr_pos_list[i]\n",
    "                if strand_list[i] == 'p':\n",
    "                    records.append(SeqRecord(record.seq[pos-10:pos+11].transcribe()))\n",
    "                elif strand_list[i] == 'n':\n",
    "                    records.append(SeqRecord(record.seq[pos-10:pos+11].transcribe().reverse_complement()))\n",
    "                SeqIO.write(records,fw,\"fasta\")\n",
    "                \n",
    "                \n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "atmospheric-retro",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-07-02 14:43:14.208272\n",
      "2021-07-02 14:43:29.213786\n",
      "chr1 end\n",
      "2021-07-02 14:43:44.242462\n",
      "chr2 end\n",
      "2021-07-02 14:43:59.392171\n",
      "chr3 end\n",
      "2021-07-02 14:44:14.575880\n",
      "chr4 end\n",
      "2021-07-02 14:44:29.808333\n",
      "chr5 end\n",
      "2021-07-02 14:44:45.286344\n",
      "chr6 end\n",
      "2021-07-02 14:45:00.724582\n",
      "chr7 end\n",
      "2021-07-02 14:45:16.026852\n",
      "chr8 end\n",
      "2021-07-02 14:45:31.329946\n",
      "chr9 end\n",
      "2021-07-02 14:45:46.632082\n",
      "chr10 end\n",
      "2021-07-02 14:46:01.892546\n",
      "chr11 end\n",
      "2021-07-02 14:46:17.107687\n",
      "chr12 end\n",
      "2021-07-02 14:46:32.323997\n",
      "chr13 end\n",
      "2021-07-02 14:46:47.492159\n",
      "chr14 end\n",
      "2021-07-02 14:47:02.731923\n",
      "chr15 end\n",
      "2021-07-02 14:47:17.891189\n",
      "chr16 end\n",
      "2021-07-02 14:47:33.061258\n",
      "chr17 end\n",
      "2021-07-02 14:47:48.212053\n",
      "chr18 end\n",
      "2021-07-02 14:48:03.282882\n",
      "chr19 end\n",
      "2021-07-02 14:48:18.408262\n",
      "chrX end\n",
      "2021-07-02 14:48:33.570457\n",
      "chrY end\n",
      "2021-07-02 14:48:48.771271\n",
      "chrM end\n",
      "2021-07-02 14:48:48.772602\n",
      "GL456210.1 end\n",
      "2021-07-02 14:48:48.773781\n",
      "GL456211.1 end\n",
      "2021-07-02 14:48:48.774959\n",
      "GL456212.1 end\n",
      "2021-07-02 14:48:48.776626\n",
      "GL456219.1 end\n",
      "2021-07-02 14:48:48.777802\n",
      "GL456221.1 end\n",
      "2021-07-02 14:49:03.994052\n",
      "GL456233.2 end\n",
      "2021-07-02 14:49:03.995499\n",
      "GL456239.1 end\n",
      "2021-07-02 14:49:03.996694\n",
      "GL456354.1 end\n",
      "2021-07-02 14:49:03.997868\n",
      "GL456359.1 end\n",
      "2021-07-02 14:49:03.999033\n",
      "GL456360.1 end\n",
      "2021-07-02 14:49:04.001119\n",
      "GL456366.1 end\n",
      "2021-07-02 14:49:19.207767\n",
      "GL456367.1 end\n",
      "2021-07-02 14:49:19.209087\n",
      "GL456368.1 end\n",
      "2021-07-02 14:49:19.210270\n",
      "GL456370.1 end\n",
      "2021-07-02 14:49:19.211550\n",
      "GL456372.1 end\n",
      "2021-07-02 14:49:34.435468\n",
      "GL456378.1 end\n",
      "2021-07-02 14:49:34.436896\n",
      "GL456379.1 end\n",
      "2021-07-02 14:49:34.438078\n",
      "GL456381.1 end\n",
      "2021-07-02 14:49:49.629782\n",
      "GL456382.1 end\n",
      "2021-07-02 14:49:49.631111\n",
      "GL456383.1 end\n",
      "2021-07-02 14:49:49.632296\n",
      "GL456385.1 end\n",
      "2021-07-02 14:49:49.633470\n",
      "GL456387.1 end\n",
      "2021-07-02 14:49:49.634657\n",
      "GL456389.1 end\n",
      "2021-07-02 14:49:49.635843\n",
      "GL456390.1 end\n",
      "2021-07-02 14:49:49.637151\n",
      "GL456392.1 end\n",
      "2021-07-02 14:49:49.638321\n",
      "GL456394.1 end\n",
      "2021-07-02 14:49:49.639487\n",
      "GL456396.1 end\n",
      "2021-07-02 14:49:49.640659\n",
      "JH584295.1 end\n",
      "2021-07-02 14:49:49.641827\n",
      "JH584296.1 end\n",
      "2021-07-02 14:49:49.642993\n",
      "JH584297.1 end\n",
      "2021-07-02 14:49:49.644165\n",
      "JH584298.1 end\n",
      "2021-07-02 14:49:49.667800\n",
      "JH584299.1 end\n",
      "2021-07-02 14:49:49.668988\n",
      "JH584300.1 end\n",
      "2021-07-02 14:49:49.670174\n",
      "JH584301.1 end\n",
      "2021-07-02 14:49:49.671344\n",
      "JH584302.1 end\n",
      "2021-07-02 14:49:49.672528\n",
      "JH584303.1 end\n",
      "2021-07-02 14:50:04.923595\n",
      "JH584304.1 end\n",
      "2021-07-02 14:50:04.924893\n",
      "MU069434.1 end\n",
      "2021-07-02 14:50:20.131533\n",
      "MU069435.1 end\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "#run the writeFasta function for each chromosome. in order to do this, go through list of positions with top entropy and separate them into distinct lists of specific chromosomes. \n",
    "with open (\"./forweblogo.fa\",\"w\") as fw:\n",
    "    print(datetime.datetime.now())\n",
    "    for chr in chr_list:\n",
    "        chr_pos_list = []\n",
    "        strand_list = []\n",
    "        for i in range(len(top_chr_mark_list)):\n",
    "            if top_chr_mark_list[i][1:] == chr:\n",
    "                chr_pos_list.append(complete_pos_arr[i])\n",
    "                strand_list.append(str(top_chr_mark_list[i])[0])\n",
    "        if chr_pos_list:  \n",
    "            writeFasta(chr, chr_pos_list, strand_list)\n",
    "        print(datetime.datetime.now())\n",
    "        print('%s end'%chr)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "innocent-germany",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
